### PR TITLE
interface inheritance examples

### LINKED ISSUE #3928: .NET examples showing interface composition to model multiple inheritance for base agents
### What feature would you like to be added?

right now the base agent examples are using .NET single inheritance vs composition.  need to show how to build base agents as interfaces. 

### Why is this needed?

this is a better pattern.

### PR DESCRIPTION
**Pull request by rysweet**
changes to AgentBase and HostBuilderExtensions to enable leveraging handlers from composition (interfaces) vs inheritance... 

closes #3928 
is related to #3925 
## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: dotnet/src/Microsoft.AutoGen/Agents/Agents/IOAgent/ConsoleAgent/IHandleConsole.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+using Microsoft.AutoGen.Abstractions;
+
+namespace Microsoft.AutoGen.Agents
+{
+    public interface IHandleConsole : IHandle<Output>, IHandle<Input>
+    {
+        string Route { get; }
+        AgentId AgentId { get; }
+        ValueTask PublishEvent(CloudEvent item);
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Agents/Agents/IOAgent/ConsoleAgent/IHandleConsole.cs:9]: Still need those methods? I thought they are already defined in `IAgentBase` interface

File: dotnet/src/Microsoft.AutoGen/Agents/HostBuilderExtensions.cs [hunk 2]
```diff
@@ -71,7 +71,52 @@ public static AgentApplicationBuilder AddAgentWorker(this IHostApplicationBuilde
                                                   .Where(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IHandle<>))
                                                   .Select(i => (GetMessageDescriptor(i.GetGenericArguments().First())?.FullName ?? "")).ToHashSet()))
                                     .ToDictionary(item => item.t, item => item.Item2);
+            // if the assembly contains any interfaces of type IHandler, then add all the methods of the interface to the eventsMap
+            var handlersMap = AppDomain.CurrentDomain.GetAssemblies()
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Agents/HostBuilderExtensions.cs:75]: That grabs all `AgentBase` implementations. What if user just wants to register partial of agents.

File: dotnet/src/Microsoft.AutoGen/Agents/Agents/IOAgent/ConsoleAgent/IHandleConsole.cs (hunk 1 above)
Comment (rysweet) [reply dotnet/src/Microsoft.AutoGen/Agents/Agents/IOAgent/ConsoleAgent/IHandleConsole.cs:9]: yes, needed for the publishevent to function from here - consider this a :partial agent. opportunity for refactor to a smaller common base or to have this interface inherit from IAgentBase

File: dotnet/src/Microsoft.AutoGen/Agents/HostBuilderExtensions.cs (hunk 2 above)
Comment (rysweet) [reply dotnet/src/Microsoft.AutoGen/Agents/HostBuilderExtensions.cs:75]: There are alternate signatures / builder methods for adding a single agent or a user defined set of AgentTypes.

**Review by LittleLittleCloud — APPROVED** [review]

**Pull request merged**